File: tool/sdk_impl/java/dart/_runtime/base/DartMap.java
// Copyright 2016, the Dart project authors.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dart._runtime.base;

import java.util.Collection;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;

import dart._runtime.types.simple.TypeEnvironment;

public interface DartMap<K, V> 
  extends Map<K, V>, dart.core.Map_interface<K, V> {

  public static dart.core.Map_interface factory$(TypeEnvironment env) {
    return null;
  }

  // TODO(springerm): Maybe this should be called LinkedHashMap?
  public class Generic<K, V> extends DartObject implements DartMap<K, V> {
    Class<K> keyType;

    Class<V> valueType;

    Map<K, V> map;

    public static <K, V> Generic<K, V> newInstance(
      Class<K> keyType, Class<V> valueType) {
      return new Generic<K, V>(keyType, valueType);
    }

    public Generic(Class<K> keyType, Class<V> valueType) {
      this.keyType = keyType;
      this.valueType = valueType;
      this.map = new HashMap<K, V>();
    }


    // --- Methods defined in Map ---

    public boolean containsValue(Object value) {
      return map.containsValue(value);
    }

    public boolean containsKey(Object key) {
      return map.containsKey(key);
    }

    public V operatorAt(Object key) {
      return map.get(key);
    }

    public void operatorAtPut(K key, V value) {
      map.put(key, value);
    }

    // TODO(springerm): Implement putIfAbsent
    // TODO(springerm): Implement addAll

    public V remove(Object key) {
      return map.remove(key);
    }

    public void clear() {
      map.clear();
    }

    // TODO(springerm): Implement forEach
    // TODO(springerm): Implement getKeys
    // TODO(springerm): Implement getValues

    public int getLength() {
      return map.size();
    }

    public boolean getIsEmpty() {
      return map.isEmpty();
    }

    public boolean getIsNotEmpty() {
      return !map.isEmpty();
    }


    // --- Methods defined in Object ---

    public int getHashCode() {
      return map.hashCode();
    }


    // --- Additional methods defined in java.util.Map ---

    public Set<Map.Entry<K, V>> entrySet() {
      return map.entrySet();
    }

    public V get(Object key) {
      return map.get(key);
    }
    
    public int hashCode() {
      return map.hashCode();
    }

    public boolean isEmpty() {
      return map.isEmpty();
    }

    public Set<K> keySet() {
      return map.keySet();
    }

    public V put(K key, V value) {
      operatorAtPut(key, value);
      return value;
    }

    public void putAll(Map<? extends K, ? extends V> m) {
      map.putAll(m);
    }

    public int size() {
      return map.size();
    }

    public Collection<V> values() {
      return map.values();
    }
  }

}

File: tool/input_sdk/internal/dart_map.dart
part of dart._internal;
 
class LinkedHashMap<K, V> implements Map<K, V> {
  List<K> _keys = new List<K>();
  List<V> _values = new List<V>();

  LinkedHashMap();

  factory LinkedHashMap.newInstance() {
    return new LinkedHashMap<K, V>();
  }

  bool containsValue(Object value) {
    return _values.contains(value);
  }

  bool containsKey(Object key) {
    return _keys.contains(key);
  }

  V operator [](Object key) {
    for (int i = 0; i < _keys.length; i++) {
      if (_keys[i] == key) {
        return _values[i];
      }
    }

    throw "Key not found";
  }

  void operator []=(K key, V value) {
    for (int i = 0; i < _keys.length; i++) {
      if (_keys[i] == key) {
        _values[i] = value;
        return;
      }
    }

    _keys.add(key);
    _values.add(value);
  }

  V remove(Object key) {
    for (int i = 0; i < _keys.length; i++) {
      if (_keys[i] == key) {
        _keys.removeAt(i);
        return _values.removeAt(i);
      }
    }

    throw "Key not found";
  }

  void clear() {
    _keys.clear();
    _values.clear();
  }

  int get length => _keys.length;

  bool get isEmpty => _keys.isEmpty;

  bool get isNotEmpty => _keys.isNotEmpty;

  Iterable<K> get keys => _keys;

  Iterable<V> get values => _values;
}